dom/events/scrolling/wheel-event-groups.html

```diff
@@ -0,0 +1,142 @@
+});
+
+let inner_wheel_event = false;
+innerDiv.addEventListener("wheel", (e) => {
+  e.stopPropagation();
+  inner_wheel_event = true;
+});
+
+function runTest() {
+  promise_test(async (t) => {
+    // Start a wheel scroll above the scrollable child element, and
+    // scroll over the child element to ensure that wheel events
+    // part of this transaction have the root element as their target.
+    await new test_driver.Actions()
+      .addWheel("wheel1")
+      .scroll(40, 2, 0, 20)
+      .scroll(40, 2, 0, 100)
+      .pause(1)
+      .scroll(40, 2, 0, 50)
+      .pause(500)
+      .send();
+
+    assert_equals(inner_wheel_event, false, "Wheel event was not fired to child");
+    assert_equals(scrollable_wheel_event, false, "Wheel event was not fired to scrollable child");
+  }, 'Wheel events are grouped and are not sent to a child element');
+
+  promise_test(async (t) => {
+    document.scrollingElement.scrollTo({top: 0, lef: 0});
+    scrollable_wheel_event = false;
+    inner_wheel_event = false;
+
+    await waitForCompositorCommit();
+
+    // Scroll in the child and ensure that wheel events in the group are not
+    // directly fired to the scrollable element.
+    await new test_driver.Actions()
+      .addWheel("wheel1")
+      .scroll(40, 2, 0, 20, {origin: innerDiv})
+      .scroll(40, 2, 0, 100, {origin: innerDiv})
+      .pause(1)
+      .scroll(40, 2, 0, 50, {origin: innerDiv})
+      .pause(500)
+      .send();
+
+    assert_equals(inner_wheel_event, true, "Wheel event was fired to inner div");
+    assert_equals(scrollable_wheel_event, false, "Wheel event was not fired to scrollable child");
+  }, 'Wheel events are not bound to a scroll frame');
+
+  promise_test(async (t) => {
+    document.scrollingElement.scrollTo({top: 0, left: 0});
+    scrollableDiv.scrollTo({top: 0, left: 0});
+    scrollable_wheel_event = false;
+    inner_wheel_event = false;
+
+    await waitForCompositorCommit();
+
+    let first_spacer_wheel_event = false;
+    firstSpacer.addEventListener("wheel", (e) => {
+      e.stopPropagation();
+      first_spacer_wheel_event = true;
+    });
+
+    let second_spacer_wheel_event = false;
+    secondSpacer.addEventListener("wheel", (e) => {
+      e.stopPropagation();
+      second_spacer_wheel_event = true;
+    });
+
+    // Scroll past the boundary of the original element to ensure all events in
+    // group have the same target.
+    await new test_driver.Actions()
+      .addWheel("wheel1")
+      .scroll(40, 2, 0, 20, {origin: firstSpacer})
+      .scroll(40, 2, 0, 100, {origin: firstSpacer})
+      .pause(1)
+      .scroll(40, 2, 0, 50, {origin: firstSpacer})
+      .pause(500)
+      .send();
+
+    assert_equals(first_spacer_wheel_event, true, "Wheel event was fired to the first spacer");
+    assert_equals(second_spacer_wheel_event, false, "Wheel event should not be fired to the second spacer");
+    assert_equals(inner_wheel_event, false, "Wheel event should not be fired to inner div");
```

Comment:
At least under current UI Events spec, one of these should be `true` if the last `.scroll` call emulates a wheel operation on `secondSpacer` or `innerDiv` because both of them are in same scrollable element.

Reply:
This is also my understanding of the current spec. This is what I was trying to describe in https://github.com/w3c/uievents/issues/338, but I'm not sure if I described it well. For both manual and emulated tests, the current spec behavior does not seem to happen on WebKit or Blink.

Note that a event sent to `secondSpacer` when the scroll moves beyond the bounds of `firstSpacer` would cause user scrolls to be interrupted in the case that `secondSpacer` has a event listener that uses preventDefault. We get a steady stream of bug reports for this in the Gecko APZ component. So maybe it makes the most sense to fire the wheel event at the parent that contains both elements instead of firing it at the initial target and relying on the event bubbling up.

Reply:
Sorry for the delay to reply, but I've still not reached the conclusion yet.

Could you rewrite these tests to collect `Event.target` of `wheel` events with `window.addEventListener` into an array and compare each array item with what you expect? Then, what's expected here becomes easier to understand. I.e.,

```
let wheelEventTargets = [];
window.addEventListener("wheel", event => {
  wheelEventTargets.push(event.target);
}, {passive: true});
...
function runTest() {
  wheelEventTargets = [];
  ...
  assert_equal(
    document.getElementById("..."),
    wheelEventTargets[0],
    "..."
  );
  assert_equal(
    document.getElementById("..."),
    wheelEventTargets[1],
    "..."
  );
}
```
(I guess and assume that the number of `wheel` events is not different between browsers in what this test does.)

Reply:
> Sorry for the delay to reply, but I've still not reached the conclusion yet.

No problem! This is a bit of a complex topic that (I think) doesn't currently follow the spec.

> Could you rewrite these tests to collect Event.target of wheel events with window.addEventListener into an array and compare each array item with what you expect?

Thanks for the feedback! Yeah, that makes sense... I think I can rework the test to work like this. Some of the way the tests are architected  is an artifact of me trying to figure out how non-firefox browsers handle wheel event group targetting.